compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -175,28 +229,71 @@ struct GPUSetEncodingOpLoweringConversion
       assert(packedShape == targetShape);
     }
 
-    // TODO(lialan): create expand_shape. Take LHS as an example:
-    // 16x4xf32 -> 16x1x4x1. Because the vector size used in the intrinsic is
-    // 1x1.
-    // For C-Matrix (i.e., ACC), it is 16x16xf32 -> 4x4x16x1xf32. Because the
-    // vector size is 4x1.
-
-    // TODO(lialan): create linalg.transpose op.
-    // LHS: 16x1x4x1 -> 4x16x1x1 (perm = [2, 0, 3, 1])
-    // ACC: 4x4x16x1 -> 4x16x4x1 (perm = [0, 2, 1, 3])
-
-    // TODO(hanchung): We want to make the shape consistent, so we need to
-    // collpase and expand the shape. This is the shape we materialize for Flow
-    // and HAL ops.
-    // 1. Create tensor.collapse_shape.
-    //    LHS: 4x16x1x1 -> 64
-    //    ACC: 4x16x4x1 -> 256
-    // 2. Create tensor.expand_shape to recover the shape (i.e., innerTiles).
-    //    LHS: 64 -> 16x4 (innerTiles[0]xinnerTiles[1])
-    //    ACC: 256 -> 16x16 (innerTiles[0]xinnerTiles[1])
-
-    // TODO(lialan): Replace the op with the tensor.expand_shape op.
-    rewriter.replaceOp(encodingOp, packOp->getResult());
+    // Create expand_shape, to tile the inner most two dimensions.
+    auto sourceShape = packOp->getResult().getType().getShape();
+    assert(intrinsicVectorShape.size() == 2); // TODO: relax this
+    auto iT1 = intrinsicVectorShape[0];
+    auto iT2 = intrinsicVectorShape[1];
+    auto oT1 = sourceShape[2] / iT1;
+    auto oT2 = sourceShape[3] / iT2;
+    SmallVector<int64_t> expandShapeShape = {
+        sourceShape[0], sourceShape[1], oT1, iT1, oT2, iT2};
+    assert(expandShapeShape.size() == 6);
+    auto expandShapeType = RankedTensorType::get(
+        expandShapeShape, encodingOp.getSourceType().getElementType());
+
+    SmallVector<ReassociationIndices> reassociationMap = {
+        {0}, {1}, {2, 3}, {4, 5}};
+    auto expandShapeOp = rewriter.create<tensor::ExpandShapeOp>(
+        loc, expandShapeType, packOp->getResult(), reassociationMap);
+
+    // create linalg.transpose on expandShapeShape
+    size_t origRank = origRank = encodingOp.getSourceType().getRank();
+    SmallVector<int64_t> transposeResultDims;
+    transposeResultDims.push_back(expandShapeShape[0]);
+    transposeResultDims.push_back(expandShapeShape[1]);
+    for (int i = 0; i < transposePermutation.size(); i++) {
+      transposeResultDims.push_back(
+          expandShapeShape[origRank + transposePermutation[i]]);
+    }
```

Comment:
We don't have to construct the shape this way. Instead, we can use the `applyPermutationToVector` helper. E.g.,

```cpp
SmallVector<int64_t> transposedShape = expandShapeShape;
applyPermutationToVector(transposedShape, newTransposePermutation);
```

https://github.com/llvm/llvm-project/blob/101acff2e5df21378b5764660afc0ec84ce6a1a6/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L220-L229